### ISSUE TITLE
Lose 70KB with One Weird Trick! Doctors hate it!

### ISSUE THREAD

**Issue by robwormald** [angular/angular#5293]
Typescript emits an `__extends` helper to, well, help with extending classes.

It currently shows up a few dozen times in our codebase, turning it off in the broccoli tree saves ~70KB unminified / 44KB minified in the existing System bundles. 

option is `noEmitHelpers: true`

Same optimization in Rx saves ~20KB or so, will submit a PR over there as well. 

We'll need to include it at least once, of course, so this might be best added into our proposed "ngPolyfill" along with Reflect?

cc @IgorMinar

**Comment by IgorMinar** [comment]
We shouldn't leak these helpers outside of our bundles otherwise we could run into conflicts with other people doing the same.

What we should do instead is include helpers only once within each bundle.

IMO this can be done post beta because it doesn't change API surface, but if someone really wants to implement this before beta then I'm not going to stand in their way.

**Comment by pkozlowski-opensource** [comment]
@robwormald did you measure things with min+gzip? I was assuming that with gzip the actual gains (in terms of size) would be much smaller. 

For me the biggest pb here is that TS doesn't give us a file with all the helpers it can produce (we don't have any equivalent of traceur-runtime).

**Comment by bradlygreen** [comment]
Even if it isn't much smaller, savings in JS parse time will still be good to have.

**Comment by pkozlowski-opensource** [comment]
tl;dr; using min+gzip numbers we can see that we can save only ~2kB (< 2%) using this trick. 

Here are exact numbers on what we can save for a simple "Hello World" application (see [diff](https://github.com/pkozlowski-opensource/angular/commit/1b7846fc073b016ad6275bff897f36d3cb511150) of what had to be changed to package helpers only once):

|  | WebPack  | SystemJS |
| ------------- | ------------- | ------------- |
| Before (bytes)  | 112 045  | 161 544  |
| After (bytes)  | 110011  | 159 402  |
| Diff (bytes)  | -2 034  (~2kB) |   -2 142 (~2.1kB) |
| Diff (%)  | -1.82% |  -1.33% |

Impact on the load time is harder to see: during my naive measures I couldn't see significant and consistent improvement.

**Comment by pkozlowski-opensource** [comment]
If we want to implement this trick, here is a TODO list:
* [ ] - make sure that TypeScript [exposes helpers](https://github.com/Microsoft/TypeScript/issues/3364) (or maintain a file with helpers on our end)
* [ ] - add TS helpers to angular2-polyfills.js
* [ ] - update bundling doc
* [ ] - update https://github.com/angular/angular2-seed

This would be a **breaking change** for people that bundle things on their own and don't use `angular2-polyfills.js`

**Comment by IgorMinar** [comment]
TS is going to ship helpers as a separate npm package. There is an issue
for 1.8 by @mhegazy about this.

**Comment by mhegazy** [comment]
the issue tracking this is https://github.com/Microsoft/TypeScript/issues/3364; along with that we should have a tslib npm package that has all the known helpers along with some .d.ts for them. This will not be in TS 1.8, but we should have it out soon though. using `--noEmitHelpers` should work in the mean time.

**Comment by Hotell** [comment]
Don't know if this is still valid issue, but in the meantime, you can use https://github.com/ngParty/ts-helpers

**Comment by imcotton** [comment]
@vicb 
Since Angular v4 heading to TypeScript v2.1, I think it's good time to revisit this ticket, with combining of `noEmitHelpers` and `importHelpers` (via dependency of `tslib`) flags.

**Comment by PFight** [comment]
@vicb Right now, Im having over 60+ __extends methods in my result bundle. Vote to open this issue again. Angular is very big, why dont to make it a bit smaller from nothing?

**Comment by IgorMinar** [comment]
The duplicates should no longer be there in 4.0.0-rc.2, can you check?
